### ISSUE TITLE
Traversing arrays within objects doesn't work

### ISSUE THREAD

**Issue by coffeepots** [status-im/nim-eth-rpc#6]
The object parsing steps through the fields recursively.
Currently, when passing an array within an object (which gets passed as something like `typeDesc[array[range[0, 10], int]]` internally)

there is the following error raised:

`Error: 'range' expects one type parameter`

The error actually points back to the object field definition, though clearly that's not where the problem is.

There's no problem with seq.

**Comment by zah** [comment]
What is the context here? I'm not sure what "object parsing" refers to.

**Comment by coffeepots** [comment]
For the `on` macro, processing the below into json getter funcs fails:

```
  type MyObject* = object
    a: int
    b: array[0..10, int]
    c: float
  s.on("rpc.objparam") do(a: string, obj: MyObject):
    # rpc code
```

The processing produces getters like this within the body of the `on` transformation (not including parameter error checking):

```
var
  a = params[0].getStr
  b: MyObject
b.a = params[1]["a"].getStr
for i in 0..<params[1]["b"].len: b.b[i] = params[1]["b"][i]
b.c = params[1]["c"].getFloat
```

Whilst it works with `array[0..10, int]` as stand alone parameters, and it works for `seq[X]` in objects, it currently fails when the array is within an object because I'm currently not doing enough to distinguish the bracket expression from other typedescs. It's just a case of improving the AST parsing code for the `on` macro, not a bug in Nim or anything.

We need this in nim-eth-rpc in order to
* Allow Nim objects to be used naturally from within the `on` body
* Add checks for all json input parameters to properly reject remote calls with the wrong number/type/length of parameters
* Avoid boiler plate of loading & checking json into structured types

**Comment by coffeepots** [comment]
@yglukhov What do you mean?

I don't think you can use overloads for `on` itself to do this, for example how would it handle nested types?

**Comment by yglukhov** [comment]
I mean you could use an overloaded primitive to unpack the arguments from json.
```nim
proc fromJson(n: JsonNode, result: var int) =
  # TODO: validate...
  result = n.getInt()

proc fromJson(n: JsonNode, result: var float) =
  # TODO: validate...
  result = n.getFloat()

proc fromJson[T](n: JsonNode, result: var seq[T]) =
  # TODO: validate...
  result.setLen(n.len)
  for i in 0 ..< n.len:
    fromJson(result[i], n[i])

proc fromJson[N, T](n: JsonNode, result: var array[N, T]) =
  # TODO: validate...
  for i in 0 ..< N:
    fromJson(result[i], n[i])

proc fromJson[T: object](n: JsonNode, result: var T) = # This reads a custom object
  # TODO: validate...
  for k, v in fieldpairs(result):
    fromJson(n[k], v)

# ... etc ...

# This should be called by the macro
proc unpackArg[T](argIdx: int, argName: string, argtype: typedesc[T], args: JsonNode): T =
  assert(argName in args)
  fromJson(arg[argName], result)

# ... etc ...

```

**Comment by coffeepots** [comment]
@yglukhov I tried using overloads as above, but it ended up becoming more complicated because of edge cases when nesting. I might revisit this later and see if I can simplify it.

To be fair we probably don't need nesting more than one level according to the RPCs but it does allow us to create more complex RPCs going forward.

I've just uploaded a commit that correctly parses objects and processes arrays correctly.
